agentstack/generation/asttools.py

```diff
@@ -0,0 +1,177 @@
+"""
+
+from typing import TypeVar, Optional, Union, Iterable
+from pathlib import Path
+import ast
+import astor
+import asttokens
+from agentstack import ValidationError
+
+
+FileT = TypeVar('FileT', bound='File')
+ASTT = TypeVar('ASTT', bound=ast.AST)
+
+
+class File:
+    """
+    Parses and manipulates a Python source file with an AST.
+
+    Use it as a context manager to make and save edits:
+    ```python
+    with File(filename) as f:
+        f.edit_node_range(start, end, new_node)
+    ```
+
+    Lookups are done using the built-in `ast` module, which we only use to find
+    and read nodes in the tree.
+
+    Edits are done using string indexing on the source code, which preserves a
+    majority of the original formatting and prevents comments from being lost.
+
+    In cases where we are constructing new AST nodes, we use `astor` to render
+    the node as source code.
+    """
+
+    filename: Path
+    source: str
+    atok: asttokens.ASTTokens
+    tree: ast.Module
+
+    def __init__(self, filename: Path):
+        self.filename = filename
+        self.read()
+
+    def read(self):
+        try:
+            with open(self.filename, 'r') as f:
+                self.source = f.read()
+                self.atok = asttokens.ASTTokens(self.source, parse=True)
+                self.tree = self.atok.tree
+        except (FileNotFoundError, SyntaxError) as e:
+            raise ValidationError(f"Failed to parse {self.filename}\n{e}")
+
+    def write(self):
+        with open(self.filename, 'w', encoding='utf-8') as f:
+            f.write(self.source)
+
+    def get_node_range(self, node: ast.AST) -> tuple[int, int]:
+        """Get the string start and end indexes for a node in the source code."""
+        return self.atok.get_text_range(node)
+
+    def edit_node_range(self, start: int, end: int, node: Union[str, ast.AST]):
+        """Splice a new node or string into the source code at the given range."""
+        if isinstance(node, ast.expr):
+            module = ast.Module(body=[ast.Expr(value=node)], type_ignores=[])
+            _node = astor.to_source(module).strip()
+        else:
+            _node = node
+
+        self.source = self.source[:start] + _node + self.source[end:]
+        # In order to continue accurately modifying the AST, we need to re-parse the source.
+        self.atok = asttokens.ASTTokens(self.source, parse=True)
+
+        if self.atok.tree:
+            self.tree = self.atok.tree
+        else:
+            raise ValidationError(f"Failed to parse {self.filename} after edit")
+
+    def __enter__(self: FileT) -> FileT:
+        return self
+
+    def __exit__(self, *args):
+        self.write()
+
+
+def get_all_imports(tree: ast.Module) -> list[ast.ImportFrom]:
+    """Find all import statements in an AST."""
+    imports = []
+    for node in ast.iter_child_nodes(tree):
+        if isinstance(node, ast.ImportFrom):  # NOTE must be in format `from x import y`
```

Comment:
how can we handle `import x`?

Reply:
just a different type (`ast.Import`)

`tool_generation.py` uses this to find the imports that already exist for a tool based on the module name, so it's gotta be a `from ...`

Could update the method name to be specific. Gets messy if we return both types because they have different signatures on the other end.